src/main/java/io/pravega/connectors/flink/serialization/PravegaDeserializationSchema.java

```diff
@@ -29,6 +30,9 @@
  * <p>This adapter exposes the Pravega serializer as a Flink Deserialization schema and
  * exposes the produced type (TypeInformation) to allow Flink to configure its internal
  * serialization and persistence stack.
+ *
+ * <p>An additional method {@link #deserializeWithMetadata(EventRead)} is provided for
+ * applying the metadata in the deserialization. This method can be overriden in the extended class. </p>
  */
 public class PravegaDeserializationSchema<T> 
```

Comment:
Type `T` here is overloaded in that it expresses both the type of object expected from the Pravega stream (used by the Pravega API) and the type of object expected from the connector (with possible Metadata injected).  This means that the object being stored in the stream MUST serialize to a Java class that expects that Metadata.  For some formats (such as JSON) this may be easy but for other formats such as Java it's difficult and messy.  In fact this can be seen in the Unit tests where `IntegerWithEventPointer` is the type being "stored" in the stream.

I'm not sure if it's possible, but it feels like there should be two types, a type `T` to represent the type from the stream and a type `C` to represent the type from the connector.  In most cases these could be the same but in other cases where MetaData needs to be injected these could be different.  

Or maybe (just thinking allowed) we're talking about two completely different interfaces, i.e. a PravegaMetaDataSchema<T, C> that extends PravegaDeserializationSchema<T> and just has the method 
```
public C deserializeWithMetadata(EventRead<T> eventRead) {
```

Reply:
Yes, this was my initial thought, and I also tried with it, but finally chose this. This is also what Kafka's approach to keep the same format of the event: https://github.com/apache/flink/blob/c6d9b37a8362317cc364c132ced5ac10dd3e10f2/flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/JSONKeyValueDeserializationSchema.java.

I wrote some investigation and the reason here: https://docs.google.com/document/d/1Dp8KYA6N81869dFjhD7w3z2hXd2QkU-vS1DNkz5aRC4/edit#